pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -48,135 +110,282 @@ func (c *WebSocketGraphQLSubscriptionClient) Subscribe(ctx context.Context, opti
 	if err != nil {
 		return err
 	}
-	defer func() {
-		if isSubscribed {
-			return
-		}
-		_ = conn.Write(ctx, websocket.MessageText, []byte(fmt.Sprintf(stopMessage, "1")))
-		_ = conn.Close(websocket.StatusNormalClosure, "")
-	}()
-	if response.StatusCode != http.StatusSwitchingProtocols {
+	if upgradeResponse.StatusCode != http.StatusSwitchingProtocols {
 		return fmt.Errorf("upgrade unsuccessful")
 	}
-
 	// init + ack
 	err = conn.Write(ctx, websocket.MessageText, connectionInitMessage)
 	if err != nil {
 		return err
 	}
-	msgType, connectionAckMesage, err := conn.Read(ctx)
+	msgType, connectionAckMsg, err := conn.Read(ctx)
 	if err != nil {
 		return err
 	}
 	if msgType != websocket.MessageText {
 		return fmt.Errorf("unexpected msg type")
 	}
-	connectionAck, err := jsonparser.GetString(connectionAckMesage, "type")
+	connectionAck, err := jsonparser.GetString(connectionAckMsg, "type")
 	if err != nil {
 		return err
 	}
 	if connectionAck != "connection_ack" {
 		return fmt.Errorf("expected connection_ack, got: %s", connectionAck)
 	}
 
-	// subscribe
+	handler = newConnectionHandler(c.ctx, conn, c.readTimeout, c.log)
+	c.handlers[handlerID] = handler
 
-	graphQLBody, err := json.Marshal(options.Body)
-	if err != nil {
-		return err
+	go func(handlerID uint64) {
+		handler.startBlocking(sub)
+		c.handlersMu.Lock()
+		delete(c.handlers, handlerID)
+		c.handlersMu.Unlock()
+	}(handlerID)
+
+	return nil
+}
+
+func (c *WebSocketGraphQLSubscriptionClient) generateHandlerIDHash(options GraphQLSubscriptionOptions) (uint64, error) {
+	var (
+		xxh *xxhash.XXHash64
+		err error
+	)
+	h := c.hashPool.Get()
+	if h == nil {
```

Comment:
will it be simpler to have a new func on a pool instead of checking for nil?

Reply:
It's more or less the same.

Reply:
Having a `New` function makes it easier to understand what to expect from the sync pool. Currently, you have to look "somewhere" in the code to find out what it "creates"